### PR TITLE
Fixes #1694 : Added HasNumberOfTermsEqualsTo Classifier Rule

### LINKED ISSUE #1694: Add classifiers HasNumberOfTermsEqualTo
Add new classifier rule HasNumberOfTermsEqualTo for RatioExpressionInput Interaction

### PR DESCRIPTION
**Pull request by aggarwalpulkit596**
## Explanation

Fixes #1694 

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: domain/src/main/java/org/oppia/domain/classify/rules/ratioinput/RatioInputHasNumberOfTermsEqualsToClassifierProvider.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+import org.oppia.app.model.InteractionObject
+import org.oppia.app.model.RatioExpression
+import org.oppia.domain.classify.RuleClassifier
+import org.oppia.domain.classify.rules.GenericRuleClassifier
+import org.oppia.domain.classify.rules.RuleClassifierProvider
+import javax.inject.Inject
+
+/**
+ * Provider for a classifier that determines whether two object have equal no of terms.
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/classify/rules/ratioinput/RatioInputHasNumberOfTermsEqualsToClassifierProvider.kt]: ``...have an equal number of terms.``

File: domain/src/main/java/org/oppia/domain/classify/rules/ratioinput/RatioInputModule.kt [hunk 2]
```diff
@@ -25,4 +25,12 @@ class RatioInputModule {
   internal fun provideRatioInputIsEquivalentRuleClassifier(
     classifierProvider: RatioInputIsEquivalentRuleClassifierProvider
   ): RuleClassifier = classifierProvider.createRuleClassifier()
+
+  @Provides
+  @IntoMap
+  @StringKey("IsEquivalent")
+  @RatioExpressionInputRules
+  internal fun provideRatioInputHasNumberOfTermsEqualsTORuleClassifier(
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/classify/rules/ratioinput/RatioInputModule.kt]:
```suggestion
  internal fun provideRatioInputHasNumberOfTermsEqualsToRuleClassifier(
```

File: domain/src/main/java/org/oppia/domain/classify/rules/ratioinput/RatioInputModule.kt [hunk 3]
```diff
@@ -25,4 +25,12 @@ class RatioInputModule {
   internal fun provideRatioInputIsEquivalentRuleClassifier(
     classifierProvider: RatioInputIsEquivalentRuleClassifierProvider
   ): RuleClassifier = classifierProvider.createRuleClassifier()
+
+  @Provides
+  @IntoMap
+  @StringKey("IsEquivalent")
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/classify/rules/ratioinput/RatioInputModule.kt]: Wrong key? I'm surprised this compiles--I'd expect Dagger to complain about the same key being used for two bindings.

File: domain/src/test/java/org/oppia/domain/classify/rules/ratioinput/RatioInputHasNumberOfTermsEqualsToClassifierProviderTest.kt [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+import kotlin.test.fail
+
+/** Tests for [RatioInputHasNumberOfTermsEqualsToClassifierProvider]. */
+@RunWith(AndroidJUnit4::class)
+@LooperMode(LooperMode.Mode.PAUSED)
+@Config(manifest = Config.NONE)
+class RatioInputHasNumberOfTermsEqualsToClassifierProviderTest {
+  private val NON_NEGATIVE_VALUE_3 = createNonNegativeInt(value = 3)
+  private val NON_NEGATIVE_VALUE_4 = createNonNegativeInt(value = 4)
+  private val ITEM_RATIO_3TERMS = listOf(1, 2, 3)
```
Comment (BenHenning) [inline domain/src/test/java/org/oppia/domain/classify/rules/ratioinput/RatioInputHasNumberOfTermsEqualsToClassifierProviderTest.kt]: ``RATIO_WITH_THREE_TERMS`` may read a bit better. ``3TERMS`` tripped me up when reading it.

File: domain/src/test/java/org/oppia/domain/classify/rules/ratioinput/RatioInputHasNumberOfTermsEqualsToClassifierProviderTest.kt [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+      hasNumberOfTermsEqualsToClassifierProvider.matches(
+        answer = createRatio(ITEM_RATIO_3TERMS),
+        inputs = inputs
+      )
+
+    assertThat(matches).isTrue()
+  }
+
+  @Test
+  fun testAnswer_testRatio_ratio4Terms_bothValuesDoNoMatch() {
```
Comment (BenHenning) [inline domain/src/test/java/org/oppia/domain/classify/rules/ratioinput/RatioInputHasNumberOfTermsEqualsToClassifierProviderTest.kt]: This reads slightly backwards to me. The 'ratio4Terms' reads as though it applies to the ratio being inputted, not the input value. Perhaps: ``testAnswer_withThreeTerms_expectedFour_doesNotMatch``?

Ditto elsehwere.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @aggarwalpulkit596! Just had a few nits.

File: domain/src/main/java/org/oppia/domain/classify/rules/ratioinput/RatioInputHasNumberOfTermsEqualsToClassifierProvider.kt (hunk 1 above)
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/classify/rules/ratioinput/RatioInputHasNumberOfTermsEqualsToClassifierProvider.kt]: done

File: domain/src/main/java/org/oppia/domain/classify/rules/ratioinput/RatioInputModule.kt (hunk 2 above)
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/classify/rules/ratioinput/RatioInputModule.kt]: done

File: domain/src/main/java/org/oppia/domain/classify/rules/ratioinput/RatioInputModule.kt (hunk 3 above)
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/classify/rules/ratioinput/RatioInputModule.kt]: oh I did this by mistake but nevermind fixed

File: domain/src/test/java/org/oppia/domain/classify/rules/ratioinput/RatioInputHasNumberOfTermsEqualsToClassifierProviderTest.kt (hunk 5 above)
Comment (aggarwalpulkit596) [reply domain/src/test/java/org/oppia/domain/classify/rules/ratioinput/RatioInputHasNumberOfTermsEqualsToClassifierProviderTest.kt]: done

**Review by BenHenning — APPROVED** [review]
Thanks @aggarwalpulkit596, LGTM!

**Pull request merged**